### ISSUE TITLE
this.event overwritten for synchronous emits within listeners (wildcard / onAny)

### ISSUE THREAD

**Issue by craig0990** [asyncly/EventEmitter2#158]
Note: sorry, I appreciate the bug title isn't brilliant, I couldn't think of a better way of putting it without writing a small essay

Disclaimer: finding this bug has made me re-think the way I'm trying to model my events, so this may be a bit of an edge case, but I thought I'd raise it anyway...

First, a minimal example:
```javascript
var EventEmitter2 = require('eventemitter2').EventEmitter2;

var emitter = new EventEmitter2({
  wildcard: true
});

// Listen for any events
emitter.on('*', function(data) {
  // Store the current event name (useful later on)
  var original_this_event = this.event;

  // Prevent infinite recursion due to wildcard listener
  if (this.event === 'something') {
    // Emit a different event inside the listener
    emitter.emit('something-else', 'more test data');
  }

  // this.event has now changed
  console.log('Recieved', this.event, 'with', data, '(original event:', original_this_event, ')');
});

emitter.emit('something', 'test data');
```

Output:

```
Recieved something-else with more test data (original event: something-else )
Recieved something-else with test data (original event: something )
```

Now, emitting events inside a wildcard listener is obviously one of those things that makes you think "WTF are you doing", because it's going to be pretty easy to get yourself into an infinite loop. For my actual use-case (I'm still experimenting with different approaches) I'm _pretty certain_ I'll never fire the same event twice, because I have a slightly different version of the `if` statement above that does pretty much the same thing.

I'm happy enough to look into how to fix this and propose a PR, but I want to check and see if this is "valid bug/use-case" or "don't be stupid - that's not what this is for" first ;)

**Comment by RangerMauve** [comment]
A potential fix might be to change what the callbacks in the events get invoked with. So rather than setting `event` on `this` and doing `fn.call(this, whatever)`, you could do something like `customThis = Object.create(this); customThis.event = event; fn.call(customThis, whatever)` this would be a bit more backwards compatible, though it might cause performance issues.

A pr would be welcome if you could find a way to do it efficiently

**Comment by RangerMauve** [comment]
Should be fixed by #141. Via a breaking change to the API. :D

**Comment by craig0990** [comment]
Ah, awesome - sorry, I missed the GitHub notification for this. I eventually moved to a different way of writing my command line utility which removed event-driven functionality altogether, but I thought I'd raise the issue.

Cheers :smiley:
